File: packages/engine/src/functions/background.ts
import Docker from "dockerode";
import { getContext } from "iii-sdk";
import type { StateKV } from "../state/kv.js";
import type { EngineConfig } from "../config.js";
import { SCOPES, generateId } from "../state/schema.js";
import { execInContainer, getDocker } from "../docker/client.js";
import { validateCommand } from "../security/validate.js";
import type { Sandbox, BackgroundExec } from "../types.js";

const BG_SCOPE = "background";

export function registerBackgroundFunctions(
  sdk: any,
  kv: StateKV,
  config: EngineConfig,
) {
  sdk.registerFunction(
    { id: "cmd::background", description: "Run command in background" },
    async (input: { id: string; command: string }): Promise<BackgroundExec> => {
      const sandbox = await kv.get<Sandbox>(SCOPES.SANDBOXES, input.id);
      if (!sandbox) throw new Error(`Sandbox not found: ${input.id}`);
      if (sandbox.status !== "running")
        throw new Error(`Sandbox not running: ${sandbox.status}`);

      const execId = generateId("bg");
      const shellCmd = `(${input.command}) > /tmp/${execId}.log 2>&1`;
      const cmd = validateCommand(shellCmd);
      const container = getDocker().getContainer(`iii-sbx-${input.id}`);

      const exec = await container.exec({
        Cmd: cmd,
        AttachStdout: false,
        AttachStderr: false,
        Detach: true,
      } as Docker.ExecCreateOptions);
      await (exec as unknown as Docker.Exec).start({ Detach: true } as any);

      const bg: BackgroundExec = {
        id: execId,
        sandboxId: input.id,
        command: input.command,
        running: true,
        startedAt: Date.now(),
      };
      await kv.set(BG_SCOPE, execId, bg);
      return bg;
    },
  );

  sdk.registerFunction(
    {
      id: "cmd::background-status",
      description: "Get background command status",
    },
    async (input: { id: string }): Promise<BackgroundExec> => {
      const bg = await kv.get<BackgroundExec>(BG_SCOPE, input.id);
      if (!bg) throw new Error(`Background exec not found: ${input.id}`);
      return bg;
    },
  );

  sdk.registerFunction(
    {
      id: "cmd::background-logs",
      description: "Get background command logs",
    },
    async (input: {
      id: string;
      cursor?: number;
    }): Promise<{ output: string; cursor: number }> => {
      const bg = await kv.get<BackgroundExec>(BG_SCOPE, input.id);
      if (!bg) throw new Error(`Background exec not found: ${input.id}`);

      const container = getDocker().getContainer(`iii-sbx-${bg.sandboxId}`);
      const logFile = `/tmp/${input.id}.log`;
      const skip = input.cursor ?? 0;
      const result = await execInContainer(
        container,
        ["sh", "-c", `tail -c +${skip + 1} ${logFile} 2>/dev/null || echo ""`],
        10000,
      );
      return {
        output: result.stdout,
        cursor: skip + Buffer.byteLength(result.stdout),
      };
    },
  );

  sdk.registerFunction(
    { id: "cmd::interrupt", description: "Interrupt a running command" },
    async (input: {
      id: string;
      pid?: number;
    }): Promise<{ success: boolean }> => {
      const sandbox = await kv.get<Sandbox>(SCOPES.SANDBOXES, input.id);
      if (!sandbox) throw new Error(`Sandbox not found: ${input.id}`);

      const container = getDocker().getContainer(`iii-sbx-${input.id}`);
      if (input.pid) {
        await execInContainer(
          container,
          ["kill", "-SIGINT", String(input.pid)],
          5000,
        );
      } else {
        await execInContainer(
          container,
          ["pkill", "-SIGINT", "-f", "sh -c"],
          5000,
        );
      }
      return { success: true };
    },
  );
}

File: packages/worker/src/functions/background.rs
use bollard::exec::CreateExecOptions;
use bollard::Docker;
use iii_sdk::III;
use serde_json::{json, Value};
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use crate::auth::validate_command;
use crate::config::EngineConfig;
use crate::docker::exec_in_container;
use crate::state::{generate_id, scopes, StateKV};
use crate::types::{BackgroundExec, Sandbox};

fn now_ms() -> u64 { SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_millis() as u64 }

pub fn register(bridge: &Arc<III>, dk: &Arc<Docker>, kv: &StateKV, _config: &EngineConfig) {
    // cmd::background
    {
        let kv = kv.clone(); let dk = dk.clone();
        bridge.register_function("cmd::background", move |input: Value| {
            let kv = kv.clone(); let dk = dk.clone();
            async move {
                let id = input.get("id").and_then(|v| v.as_str())
                    .ok_or_else(|| iii_sdk::IIIError::Handler("id is required".into()))?;
                let command = input.get("command").and_then(|v| v.as_str())
                    .ok_or_else(|| iii_sdk::IIIError::Handler("command is required".into()))?;
                let sandbox: Sandbox = kv.get(scopes::SANDBOXES, id).await
                    .ok_or_else(|| iii_sdk::IIIError::Handler(format!("Sandbox not found: {id}")))?;
                if sandbox.status != "running" {
                    return Err(iii_sdk::IIIError::Handler(format!("Sandbox not running: {}", sandbox.status)));
                }

                let exec_id = generate_id("bg");
                let shell_cmd = format!("({command}) > /tmp/{exec_id}.log 2>&1");
                let cmd = validate_command(&shell_cmd)
                    .map_err(|e| iii_sdk::IIIError::Handler(e))?;

                let cn = format!("iii-sbx-{id}");
                let exec = dk.create_exec(&cn, CreateExecOptions {
                    cmd: Some(cmd.iter().map(|s| s.as_str()).collect()),
                    attach_stdout: Some(false),
                    attach_stderr: Some(false),
                    ..Default::default()
                }).await.map_err(|e| iii_sdk::IIIError::Handler(format!("Exec create failed: {e}")))?;

                dk.start_exec(&exec.id, None).await
                    .map_err(|e| iii_sdk::IIIError::Handler(format!("Exec start failed: {e}")))?;

                let bg = BackgroundExec {
                    id: exec_id.clone(), sandbox_id: id.to_string(),
                    command: command.to_string(), pid: None,
                    running: true, exit_code: None,
                    started_at: now_ms(), finished_at: None,
                };
                kv.set(scopes::BACKGROUND, &exec_id, &bg).await
                    .map_err(|e| iii_sdk::IIIError::Handler(e.to_string()))?;
                serde_json::to_value(&bg).map_err(|e| iii_sdk::IIIError::Serde(e.to_string()))
            }
        });
    }

    // cmd::background-status
    {
        let kv = kv.clone();
        bridge.register_function("cmd::background-status", move |input: Value| {
            let kv = kv.clone();
            async move {
                let id = input.get("id").and_then(|v| v.as_str())
                    .ok_or_else(|| iii_sdk::IIIError::Handler("id is required".into()))?;
                let bg: BackgroundExec = kv.get(scopes::BACKGROUND, id).await
                    .ok_or_else(|| iii_sdk::IIIError::Handler(format!("Background exec not found: {id}")))?;
                serde_json::to_value(&bg).map_err(|e| iii_sdk::IIIError::Serde(e.to_string()))
            }
        });
    }

    // cmd::background-logs
    {
        let kv = kv.clone(); let dk = dk.clone();
        bridge.register_function("cmd::background-logs", move |input: Value| {
            let kv = kv.clone(); let dk = dk.clone();
            async move {
                let id = input.get("id").and_then(|v| v.as_str())
                    .ok_or_else(|| iii_sdk::IIIError::Handler("id is required".into()))?;
                let cursor = input.get("cursor").and_then(|v| v.as_u64()).unwrap_or(0);

                let bg: BackgroundExec = kv.get(scopes::BACKGROUND, id).await
                    .ok_or_else(|| iii_sdk::IIIError::Handler(format!("Background exec not found: {id}")))?;

                let cn = format!("iii-sbx-{}", bg.sandbox_id);
                let log_file = format!("/tmp/{id}.log");
                let cmd = vec!["sh".into(), "-c".into(), format!("tail -c +{} {} 2>/dev/null || echo \"\"", cursor + 1, log_file)];
                let result = exec_in_container(&dk, &cn, &cmd, 10000).await
                    .map_err(|e| iii_sdk::IIIError::Handler(e))?;
                let new_cursor = cursor + result.stdout.len() as u64;
                Ok(json!({ "output": result.stdout, "cursor": new_cursor }))
            }
        });
    }

    // cmd::interrupt
    {
        let kv = kv.clone(); let dk = dk.clone();
        bridge.register_function("cmd::interrupt", move |input: Value| {
            let kv = kv.clone(); let dk = dk.clone();
            async move {
                let id = input.get("id").and_then(|v| v.as_str())
                    .ok_or_else(|| iii_sdk::IIIError::Handler("id is required".into()))?;
                let _sandbox: Sandbox = kv.get(scopes::SANDBOXES, id).await
                    .ok_or_else(|| iii_sdk::IIIError::Handler(format!("Sandbox not found: {id}")))?;
                let cn = format!("iii-sbx-{id}");
                let pid = input.get("pid").and_then(|v| v.as_u64());
                let cmd = if let Some(p) = pid {
                    vec!["kill".into(), "-SIGINT".into(), p.to_string()]
                } else {
                    vec!["pkill".into(), "-SIGINT".into(), "-f".into(), "sh -c".into()]
                };
                exec_in_container(&dk, &cn, &cmd, 5000).await
                    .map_err(|e| iii_sdk::IIIError::Handler(e))?;
                Ok(json!({ "success": true }))
            }
        });
    }
}
